### PR TITLE
[Inductor] Filter triton configs by launch geometry in triton heuristics

### LINKED ISSUE #182928: [Inductor] Filter extremely bad triton configs by device information for triton heuristics
### 🚀 The feature, motivation and pitch

The current workflow for the triton heuristics is to generate a series of triton configs and then run autotune to a best config. Usually these config generated are based on well known practices, and at the same time try to cover as many cases as possible because autotune is a onetime effort and it doesn't seem to hurt much to evaluate more configs. With max_autotune, it also tend to add or keep more configs to bench.

This workflow is simple and straight forward. But we saw the config included lacks the consideration of the device capabilities such as the number CTAs especially when the GPU in the market becomes more diverse: low end GPU may have a few CTAs per GPU, while high end GPU have hundreds. For example, for the following case:

```
import torch
import time

start_time = time.perf_counter()

class TestModule(torch.nn.Module):
    # Input tensors that are generated randomly
    def __init__(self, sel_device, training):
        torch.manual_seed(777)
        super(TestModule, self).__init__()
        self.inp = torch.empty_strided(size=[16, 4096, 1024], stride=[4194304, 1024, 1], dtype=torch.bfloat16, device='cpu').uniform_(-1, 1).to(device=sel_device).requires_grad_(training)

    def forward(self):
        r = torch.sum(self.inp, [0], False) # traced line: 1010
        return r

device = "xpu"
m = TestModule(device, False)
m = torch.compile(m)     

out = m()
out = out.to("cpu")

end_time = time.perf_counter()
print("Total time:", str(end_time - start_time))
```
Inductor persistent_reduction will generate and auto tune 4 XBLOCK configurations with the reduction size = 16:
```
1. {'XBLOCK': 1}
2. {'XBLOCK': 8}  
3. {'XBLOCK': 32}  
4. {'XBLOCK': 128} 
```
These will bench 4 grid configurations:
```
1. {'XBLOCK': 1} -> grid (4194304, 1, 1)
2. {'XBLOCK': 8}  -> grid (524288, 1, 1)
3. {'XBLOCK': 32}  -> grid (131072, 1, 1)
4. {'XBLOCK': 128} -> grid (32768, 1, 1)
```
Extreme cases like 4194304 thread blocks /work groups with reduction on 16 elements is obviously bad and will mostly perform the worst. We will spend much of the time benchmarking some very bad config without a chance to win. What's more, when we running this config in weak GPU or in a simulator, this kind of extreme case will take much of the time and make the benchmark time unacceptable.

One method to deal with this is to avoid including such a config at the beginning by applying the knowledge of the problem size and the device capabilities. But this will make the config logic too complex and dynamic considering we already have a lot of cases to handle for pointwise, reduction, persistent reduction and so on. The existing way doing limited check and adding configs is still ideal for simplicity.

The suggest approach is to simply add the capability to perform a scrub process which filter out the extreme triton config by device information after all configs are created.  This makes the logic of filtering centralized and not intrusive for the config generation logic. And this filtering config by device information can be controlled by a flag so that user can turn on or off.

@ezyang @EikanWang 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jerrychenhf**
Implements the idea proposed in #182928:
1. Implement the function to filter out extremely bad configs based on the grid size, cta work and the number of ctas. Keep the smallest grid size if all the configs are filtered.
2. Use the function to filter configs at the end of pointwise, reduction, persistent_reduction, cooperative_reduction, split_scan.
3. Add a flag (torch._inductor.config.triton.filter_configs_by_launch_geometry or TORCHINDUCTOR_TRITON_FILTER_CONFIGS_BY_LAUNCH_GEOMETRY env) in the triton config to turn on or off the feature. By default turn off for now. 

Fixes #182928

Authored with Claude.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by etaf** [comment]
@claude review PR

File: torch/_inductor/runtime/triton_heuristics.py [hunk 1]
```diff
@@ -4377,6 +4383,129 @@ def _pick_config():
...
+    inductor_meta: dict[str, Any],
+    triton_meta: dict[str, Any],
+    configs: list[Config],
+) -> list[Config]:
+    """
+    Filter extremely poor launch configurations based on launch geometry and
+    device parallelism, while preserving at least one candidate.
+    """
+    device = triton_meta["device"]
+    if (
```
Comment (etaf) [inline torch/_inductor/runtime/triton_heuristics.py:4386]: Currently we only enable this config for XPU, because we have not tested on CUDA/ROCM.

**Review by jansel — CHANGES_REQUESTED** [review]
fix merge conflicts

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Review by jansel — CHANGES_REQUESTED** [review]
fix merge conflicts

**Comment by jerrychenhf** [comment]
@jansel My dev branch was synced with pytorch:main by merge which causing your rebase command failure.
I will try to clean my dev branch for rebase.

**Comment by jerrychenhf** [comment]
@jansel It's now clean for rebase. Sorry for these round trips I caused.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@jerrychenhf have you done any benchmarking of this?  How well does the filtering work?

**Comment by jerrychenhf** [comment]
> @jerrychenhf have you done any benchmarking of this? How well does the filtering work?

@jansel I didn't test on a real GPU hardware. I notice this improvement from running some tests on GPU simulator that a reduce op (like torch.sum) reducing on a small dim with large size in other dimensions. This result HUGE grid size with small reduce elements when XBLOCK = 1 (a config in reduce persistent).  

The extremely bad config in benchmarking takes 70% of the case running on simulator. Though improvement will not be that much on real GPU hardware, considering there are some low end GPUs, it is still a good practice to filter out the worst configs (takes longest time compare with other configs) and benchmark less configs.

**Comment by jansel** [comment]
I think we need to test this on real hardware/models.

**Comment by etaf** [comment]
Hi, @jerrychenhf , would you like to move the filter to `XPUReductionHeuristic` in https://github.com/pytorch/pytorch/blob/408ee92f1a6ed1d761f9ab880947c44bade9f15f/torch/_inductor/heuristics/triton_codegen/reduction.py#L721 ? 
And I think we need to remove the config since it's only for xpu.

**Review by etaf — CHANGES_REQUESTED** [review]
